### PR TITLE
Synchronize buffer with server before formatting

### PR DESCRIPTION
**Pull request by mkcms**
This patch will give user the option to synchronize buffer contents with server before sending `TextDocument/formatting` request.

### CODE REVIEW

**Comment by joaotavora** [comment]
What happens exactly if the buffer contents aren't synchronized?

**Comment by mkcms** [comment]
> What happens exactly if the buffer contents aren't synchronized?

If you type something and immediately try to format the buffer, the changes can be lost, unless the timer in `eglot--after-change` gets a chance to run.

File: eglot.el [hunk 1]
```diff
@@ -1388,27 +1388,45 @@ DUMMY is ignored."
...
+  (eglot--widening
+   (let* ((server (eglot--current-server-or-lose))
+          (resp (eglot--request
+                 server
+                 :textDocument/formatting
+                 (list :textDocument (eglot--TextDocumentIdentifier)
+                       :options (list :tabSize tab-width
+                                      :insertSpaces
+                                      (if indent-tabs-mode :json-false t)))))
+          (changes
```
Comment (joaotavora) [inline eglot.el]: I'm dissapointed that you abandoned `eglot--apply-text-edits`, you seem to be reimplementing it. What's missing from that function?  Is it the `replace-buffer-contents` call?  If so, feel free to change `eglot--apply-text-edits` (i didn't know about `replace-buffer-contents` at all until you used it, it sounds like the right tool to use for text edits in general).

**Comment by joaotavora** [comment]
> If you type something and immediately try to format the buffer, the changes can be lost, unless the timer in eglot--after-change gets a chance to run.

Aha. But passing a non-nil `:deferred` argument to `eglot--request` is precisely designed to take care of this problem. You were doing that in the previous commit, but took it out in this one.  Can you re-add and retry?

File: eglot.el (hunk 1 above)
Comment (mkcms) [reply eglot.el]: > eglot--apply-text-edits

This function is broken when any TextEdit is empty (actually `eglot--range-region` is). I didn't want to break any existing behavior.

Comment (joaotavora) [reply eglot.el]: If it's broken, then fix it :-D

**Comment by mkcms** [comment]
> Aha. But passing a non-nil :deferred argument to eglot--request is precisely designed to take care of this problem.

Oh, it wasn't very clear for me from the docstring.  I think this can be closed then, I will make the changes in other PR.

**Pull request closed without merging**